tests/test_notify_suppliers_of_brief_withdrawal.py

```diff
@@ -0,0 +1,94 @@
+import mock
+from datetime import date
+
+from dmscripts import notify_suppliers_of_brief_withdrawal as tested_script
+
+
+WITHDRAWN_BRIEFS = {"briefs": [
+    {
+        "id": 123,
+        "withdrawnAt": "2016-01-28 16:23:50.618053",
+        "title": "Tea Drinker",
+        "frameworkFramework": "digital-outcomes-and-specialists"
+    },
+    {
+        "id": 235,
+        "withdrawnAt": "2016-01-28 08:23:50.618053",
+        "title": "Cookie Muncher",
+        "frameworkFramework": "digital-outcomes-and-specialists"
+    }
+]}
+
+
+BRIEF_RESPONSES = {"briefResponses": [
+    {"id": 4321, "respondToEmailAddress": "email@me.now"},
+    {"id": 4389, "respondToEmailAddress": "email@them.now"}
+]}
+
+
+EXPECTED_BRIEF_CONTEXT = {
+    'brief_title': "Tea Drinker",
+    'brief_link': 'https://www.preview.marketplace.team/digital-outcomes-and-specialists/opportunities/123'  # noqa
+}
+
+
+@mock.patch('dmapiclient.DataAPIClient', autospec=True)
+def test_get_brief_response_emails(data_api_client):
+    data_api_client.find_brief_responses.return_value = BRIEF_RESPONSES
+    assert tested_script.get_brief_response_emails(data_api_client, {"id": 1234}) == ["email@me.now", "email@them.now"]
+
+
+def test_create_context_for_brief():
+    assert tested_script.create_context_for_brief('preview', WITHDRAWN_BRIEFS["briefs"][0]) == EXPECTED_BRIEF_CONTEXT
+
+
+@mock.patch('dmutils.email.DMNotifyClient', autospec=True)
+@mock.patch('dmscripts.notify_suppliers_of_brief_withdrawal.create_context_for_brief')
+@mock.patch('dmscripts.notify_suppliers_of_brief_withdrawal.get_brief_response_emails')
+@mock.patch('dmapiclient.DataAPIClient', autospec=True)
+def test_main(data_api_client, get_brief_response_emails, create_context_for_brief, notify_client):
+    data_api_client.find_briefs.return_value = WITHDRAWN_BRIEFS
+    tested_script.create_context_for_brief.return_value = EXPECTED_BRIEF_CONTEXT
+    get_brief_response_emails.side_effect = [["email@me.now", "email@them.now"], []]
+    withdrawn_date = date(2016, 1, 28)
+    tested_script.main(data_api_client, notify_client, "notify_template_id", "preview", mock.Mock(), withdrawn_date)
+    assert data_api_client.find_briefs.call_args_list == [mock.call(withdrawn_on=withdrawn_date)]
+    assert create_context_for_brief.call_args_list == [mock.call("preview", WITHDRAWN_BRIEFS["briefs"][0])]
+    assert notify_client.send_email.call_args_list == [
+        mock.call("email@me.now", "notify_template_id", EXPECTED_BRIEF_CONTEXT, allow_resend=False),
+        mock.call("email@them.now", "notify_template_id", EXPECTED_BRIEF_CONTEXT, allow_resend=False)
+    ]
+
+
+@mock.patch('dmutils.email.DMNotifyClient', autospec=True)
+@mock.patch('dmapiclient.DataAPIClient', autospec=True)
+def test_main_calls_correct_client_methods(data_api_client, notify_client):
```

Comment:
I'm not sure what the difference is between this test and the one above? Could they be combined with a more descriptive test name e.g. `test_main_notifies_all_withdrawn_briefs_by_default` (to differentiate it from the `brief_id` test below?

Reply:
This test explicitly checks client methods and nothing else. The one above mocks _everything_ and checks the correct methods in the script are called.

The first is at the boundary of the methods - method within the scripts
The second is at the boundary of the script - clients

Reply:
OK that makes sense, but both tests assert the `data_api_client` and `notify_client` calls at the moment, and the first one mocks out `get_brief_response_emails` but doesn't assert its call. 

To help devs with poor memories (i.e. me) I'd rename the tests to make it clear what you're doing - maybe something like `test_main_calls_script_methods` and `test_main_calls_external_clients`?

Reply:
Done

Reply:
Cool, much better 😎.  The first test should still be asserting the calls for the mocked `get_brief_response_emails` script method though?

Reply:
Boom 👍